qiskit/circuit/quantumcircuit.py

```
@@ -4807,8 +4807,10 @@ def _unroll_param_dict(
                         f" but was assigned to {len(value)} values."
                     )
                 out.update(zip(parameter, value))
-            elif isinstance(parameter, str):
+            elif isinstance(parameter, str) and self.has_parameter(parameter):
                 out[self.get_parameter(parameter)] = value
+            elif isinstance(parameter, str) and not self.has_parameter(parameter):
+                out[Parameter(parameter)] = value
```

If the `QuantumCircuit` doesn't have the parameter, I don't think there's any benefit making a dummy `Parameter` instance - it'll never bind to anything anyway.  I think we can just not add it to the dictionary if it's not present and we're in `strict=False` mode, but we still need to throw an error if we're in `strict=True`.  We might need to thread a `strict` keyword argument down to this inner method.

---

@jakelishman @mtreinish Thanks for the suggestions. However, even if we pass `strict` to `_unroll_param_dict`, the fact that a dictionary with string keys gets passed to `target._data.assign_parameters_mapping(raw_mapping)` will remain a problem, unless we convert the keys to `Parameter` type. If the parameters are already present in the circuit, it is not a problem because the method `get_parameter` converts the keys from string type to `Parameter` type. However, if the parameters are not present in the circuit, we will have to create a `Parameter` instance out of the strings. 

Once it is passed as a dictionary with Parameter keys to the `assign_parameters` method, the rest of the logic falls in place. We don't really need to pass `strict ` to the `_unroll_param_dict` method.

---

Changed as suggested by @jakelishman. `strict` argument passed to the inner method.